src/actions/tests/breakpoints.js

```diff
@@ -24,11 +51,40 @@ const simpleMockThreadClient = {
 describe("breakpoints", () => {
   it("should add a breakpoint", async () => {
     const { dispatch, getState } = createStore(simpleMockThreadClient);
+    const loc1 = { sourceId: "a", line: 5 };
+
+    await dispatch(actions.addBreakpoint(loc1));
+    const bps = selectors.getBreakpoints(getState());
+    const bp = selectors.getBreakpoint(getState(), loc1);
+    expect(bps.size).to.be(1);
+    expect(bp.location).to.eql(loc1);
+  });
+
+  it("should not re-add a breakpoint", async () => {
+    const { dispatch, getState } = createStore(simpleMockThreadClient);
+    const loc1 = { sourceId: "a", line: 5 };
+
+    await dispatch(actions.addBreakpoint(loc1));
+    let bps = selectors.getBreakpoints(getState());
+    const bp = selectors.getBreakpoint(getState(), loc1);
+    expect(bps.size).to.be(1);
+    expect(bp.location).to.eql(loc1);
+
+    await dispatch(actions.addBreakpoint(loc1));
+    bps = selectors.getBreakpoints(getState());
+    expect(bps.size).to.be(1);
+  });
 
-    await dispatch(actions.addBreakpoint({ sourceId: "a", line: 5 }));
-    await dispatch(actions.addBreakpoint({ sourceId: "b", line: 6 }));
+  it("when the user adds a sliding breakpoint, a corresponding pending breakpoint should be added", async () => {
+    const { dispatch, getState } = createStore(slidingMockThreadClient);
```

Comment:
nitpick: I think it might be easy to break this, by not getting the right offset between slidingMockThreadClient and slidLocation. We can explicitly associate these two (or more) stub elements by using a wrapper function. Here is an example of how it might be done:

```
function generateMockThreadClient (offset = 0) {
  return { 
    setBreakpoint: (location, condition) => {
      return new Promise((resolve, reject) => {
        const actualLocation = Object.assign({}, location, {
          line: location.line + offset
        });
        resolve({ id: makeLocationId(location), actualLocation, condition });
      });
    }
  };
};

function generateOffsetThreadClient(offset, location) => {
  const slidingMockThreadClient = generateSlidingMockThreadClient(offset);
  const slidLocation = Object.assign({}, location, { line: location.line + offset });
  return { slidingMockThreadClient, slidLocation };
}

function ... () { // wherever this will be used
  const location = { sourceId: "a", line: 5 };
  const { slidingMockThreadClient, slidLocation } = generateOffsetThreadClient(2, location);
  const { dispatch, getState } = createStore(slidingMockThreadClient);
  ... // and so on
 }
```

naming needs to be improved. it might make sense to make location an array, but i think its getting ahead of ourselves